### DISCUSSION TITLE
form.Submit() seems to be hanging in hook

### DISCUSSION THREAD

**Discussion by kisamoto** [pocketbase/pocketbase#3205] [Q&A] [answered]
Previously I wrote about a [soft delete hook](https://github.com/pocketbase/pocketbase/discussions/2694) that stores a copy of the deleted record in a separate collection. 

I have since modified the hook from a Record to a Model hook (`OnModelBeforeDelete`). Hook code below.

However for some reason, when I delete a record (e.g. from the admin interface) the whole process hangs until I kill the server. After some digging I found it's hanging in the `form.Submit()` method but I don't know why. It appears that it's trying to perform the insert but failing, is there a transaction or something that's blocking the write? 

Updated hook:

```go
func SoftDelete(app core.App) hook.Handler[*core.ModelEvent] {
	return func(e *core.ModelEvent) error {

		record, ok := e.Model.(*models.Record)

		if !ok {
			return nil
		}

		// Add other collection names here where you don't want to keep a copy of the deleted record
		excludedCollections := []string{
			DeletedCollectionName,
		}

		// Skip if the record is already deleted or in excluded collections
		if slices.Contains(excludedCollections, record.Collection().Name) {
			return nil
		}

		collection, err := app.Dao().FindCollectionByNameOrId(DeletedCollectionName)
		if err != nil {
			return err
		}

		deletedRecord := models.NewRecord(collection)

		form := forms.NewRecordUpsert(app, deletedRecord)
		form.LoadData(map[string]any{
			"collection": record.Collection().Name,
			"record":     record.ColumnValueMap(),
		})

		return form.Submit()
	}
}
```

After I kill the server (mid-hang), I see a log line like this, so it is trying to insert into the `deleted` collection:

```
[16085.00ms] INSERT INTO `deleted` (`collection`, `created`, `id`, `record`, `updated`) VALUES ('test_instance', '2023-08-27 13:02:21.797Z', 'ac8rf8f0357ivwq', '{"created":"2023-08-27 12:56:13.931Z","custom_domain":"","deleted":"","display_name":"test","expired":"","id":"n5bhgxde28iisr1","slug":"test","status":"PENDING","updated":"2023-08-27 12:56:13.931Z"}', '2023-08-27 13:02:21.797Z')
```

**Comment by ganigeorgiev** [comment, marked as answer]
I suspect that it is conflicting with the internal delete transaction.

You can try to explicitly set the form Dao to the one from the event, eg. with `form.SetDao(e.Dao)` after form initialization.

If you don't need data validations, you can also skip the form entirely and directly persist the record:

```go
deletedRecord := models.NewRecord(collection)

deletedRecord.Set("collection", record.Collection().Name)
deletedRecord.Set("record", record.ColumnValueMap())

return e.Dao.SaveReord(deletedRecord)
```

**Reply by kisamoto** [reply]
Thanks for the fast response. I like to use validations wherever possible so I tried with the `form.SetDao` approach and it worked! 🎉

Many thanks and happy hacking
